aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -1,12 +1,63 @@
 #include <ATen/native/ScatterGatherShapeChecks.h>
 #include <ATen/native/DispatchStub.h>
 #include <ATen/native/TensorIterator.h>
+#include <ATen/native/TensorAdvancedIndexing.h>
 #include <ATen/Parallel.h>
 
 namespace at { namespace native {
 
 namespace {
 
+// Implement as functors since lambdas don't get optimized.
+class ReduceMultiply {
+public:
+  template <typename scalar_t>
+  void operator()(scalar_t * self_data, scalar_t * src_data) {
+    *self_data *= *src_data;
+  };
+
+  void operator()(bool * self_data, bool * src_data) {
+    *self_data = *self_data && *src_data;
+  };
+};
+ReduceMultiply reduce_multiply;
+
+class ReduceAdd {
+public:
+  template <typename scalar_t>
+  void operator()(scalar_t * self_data, scalar_t * src_data) {
+    *self_data += *src_data;
+  };
```

Comment:
```suggestion
  template <typename scalar_t>
  constexpr void operator()(scalar_t * self_data, scalar_t * src_data)  const {
    *self_data += *src_data;
  };
```
Starting from C++14 this implies inlining. Same for other functors. Could you please give it a try?

Reply:
Hmm, but it could be that `constexpr` might not work with void functions however, at least in C++11 this is the case.